app/models/spree/stock_movement_decorator.rb

```diff
@@ -0,0 +1,10 @@
+Spree::StockMovement.class_eval do
+  after_save :refresh_products_cache
```

Comment:
This is the sequence of method calls :face_with_head_bandage: up to `StockMovement` when a line item is destroyed for the problem scenario:

* `LineItem#update_inventory_before_destroy`
* `LineItem#update_inventory`, as alias calling
* `LineItem#update_inventory_with_scoping`
* `LineItem#update_inventory_without_scoping`, which calls
* Spree core `LineItem#update_inventory`
* `Spree::OrderInventory#verify`
* `Spree::OrderInventory#remove`
* `Spree::OrderInventory#remove_from_shipment`
* `Spree::InventoryUnit#destroy`
* `Spree::StockLocation#move` / `Spree::StockLocation#restock`
* Scoped `Spree::Variant#move`, which is `OpenFoodNetwork::ScopeVariantToHub::ScopeVariantToHub#move`
* `OpenFoodNetwork::ScopeVariantToHub::ScopeVariantToHub#move_stock!`
* `Spree::Variant#move` in `VariantStock`
* `Spree::StockMovement#create`

Reply:
For overridden limited stock variants, the cache refresh happens in `VariantOverride`.

Reply:
Thanks a lot for working on this Kristina!

That stack trace is for deleting line item of a completed order (Spree::OrderInventory#verify will only move stock on completed orders)

I think the basic problem scenario is the checkout flow which goes:
- Spree::Order#finalize!
- Spree::Shipment#finalize!
- Spree::Shipment#manifest_unstock
- Spree::StockLocation#unstock
- Spree::StockLocation#move where this line meets the line you described above

I think your solution is valid! I think we could alternatively do, if you prefer, refresh_products_cache inside VariantStock#move